serialise/src/Codec/Serialise/Class.hs

```diff
@@ -306,18 +308,48 @@ instance Serialise (Proxy a) where
     decode   = Proxy <$ decodeNull
 #endif
 
+-- | Is a 'Char' a UTF-16 surrogate?
+isSurrogate :: Char -> Bool
+isSurrogate c = (ord c .&. 0xd800) == 0xd800
+
 -- | @since 0.2.0.0
 instance Serialise Char where
-    encode c = encodeString (Text.singleton c)
-    decode = do t <- decodeString
-                if Text.length t == 1
-                  then return $! Text.head t
-                  else fail "expected a single char, found a string"
+    -- Here we've taken great pains to ensure that surrogate characters, which
+    -- are not representable in UTF-8 yet still admitted by Char,
+    -- round-trip properly. We scan the encoded characters during encoding
+    -- looking for surrogates; if we find any we encode the string as a
+    -- a list of code-points encoded as words. This is slow, but should be rare.
+    encode c
+      | isSurrogate c = encodeWord (fromIntegral $ ord c)
+      | otherwise     = encodeString (Text.singleton c)
+    decode = do ty <- peekTokenType
+                case ty of
+                  TypeUInt -> chr . fromIntegral <$> decodeWord
```

Comment:
In principle we should turn the use of partial `ord` here into failure in the decoder.